hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/GrpcOmTransport.java

```diff
@@ -0,0 +1,330 @@
+        grpcException = cn.newInstance(status.getDescription());
+        IOException remote = null;
+        try {
+          String cause = status.getDescription();
+          cause = cause.substring(cause.indexOf(":") + 2);
+          remote = new RemoteException(cause.substring(0, cause.indexOf(":")),
+              cause.substring(cause.indexOf(":") + 1));
+          grpcException.initCause(remote);
+        } catch (Exception e) {
+          LOG.error("cannot get cause for remote exception");
+        }
+      } else {
+        // exception generated by connection failure, gRPC
+        grpcException = ex;
+      }
+    } catch (Exception e) {
+      grpcException = new IOException(e);
+      LOG.error("error unwrapping exception from OMResponse {}");
+    }
+    return grpcException;
+  }
+
+  private boolean shouldRetry(Exception ex) {
+    boolean retry = false;
+    RetryPolicy.RetryAction action = null;
+    try {
+      action = retryPolicy.shouldRetry((Exception)ex, 0, failoverCount++, true);
+      LOG.debug("grpc failover retry action {}", action.action);
+      if (action.action == RetryPolicy.RetryAction.RetryDecision.FAIL) {
+        retry = false;
+        LOG.error("Retry request failed. " + action.reason, ex);
+      } else {
+        if (action.action == RetryPolicy.RetryAction.RetryDecision.RETRY ||
+            (action.action == RetryPolicy.RetryAction.RetryDecision
+                .FAILOVER_AND_RETRY)) {
+          if (action.delayMillis > 0) {
+            try {
+              Thread.sleep(action.delayMillis);
+            } catch (Exception e) {
+              LOG.error("Error trying sleep thread for {}", action.delayMillis);
+            }
+          }
+          // switch om host to current proxy OMNodeId
+          omFailoverProxyProvider.performFailover(null);
+          host.set(omFailoverProxyProvider
+              .getGrpcProxyAddress(
+                  omFailoverProxyProvider.getCurrentProxyOMNodeId()));
+          retry = true;
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed failover exception {}", e);
+    }
+    return retry;
+  }
+
+  // stub implementation for interface
+  @Override
+  public Text getDelegationTokenService() {
+    return new Text();
+  }
+
+  public void shutdown() {
+    for (Map.Entry<String, ManagedChannel> entry : channels.entrySet()) {
+      ManagedChannel channel = entry.getValue();
+      channel.shutdown();
+      try {
+        channel.awaitTermination(5, TimeUnit.SECONDS);
+      } catch (Exception e) {
+        LOG.error("failed to shutdown OzoneManagerServiceGrpc channel {} : {}",
+            entry.getKey(), e);
+      }
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    shutdown();
+  }
+
+  /**
+   * GrpcOmTransport configuration in Java style configuration class.
+   */
+  @ConfigGroup(prefix = "ozone.om.grpc")
+  public static final class GrpcOmTransportConfig {
+    @Config(key = "port", defaultValue = "8981",
```

Comment:
GrpcOmTransport.GrpcOmTransportConfig and GrpcOzoneManagerServer.GrpcOzoneManagerServerConfig both define the same configuration (`ozone.om.grpc.port`).  I think we only need one of them.

Reply:
We can use one "_**ozone.om.grpc.port**_", however currently it is split between the client (`GrpcOmTransportConfig`) and the server (_GrpcOzoneManagerServerConfig_).  Using one for the other in the case of the _GrpcOzoneManagerServerConfig_ for the client causes import problems, ie. _**package org.apache.hadoop.ozone.om cannot be found**_ due to it located in the OzoneManager path, unavail to the hadoop-ozone/common.

Can the port be referenced by both the `GrpcOmTransportConfig` and the `GrpcOzoneManagerServerConfig` as it is?  If not, if combined where is a good package location for the common `Grpc..Config` class?

Reply:
> _GrpcOzoneManagerServerConfig_ for the client causes import problems, ie. _**package org.apache.hadoop.ozone.om cannot be found**_ due to it located in the OzoneManager path, unavail to the hadoop-ozone/common.

`hadoop-ozone/common` contains sources shared between Ozone client and server-side components.  So the config defined there can be used in OM, no need to define another class for it.  In other words, keep `GrpcOmTransportConfig` and drop `GrpcOzoneManagerServerConfig` (until a server-only config is needed).

Reply:
Thanks @adoroszlai.  No import problem with that.  Updated to reuse the `ConfigGroup` definition for the gprc port from `hadoop-ozone/common/GrpcOmTransport` for the server as suggested.